// @openzeppelin/wizard/dist/add-pausable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPausable = addPausable;
exports.addPauseFunctions = addPauseFunctions;
const set_access_control_1 = require("./set-access-control");
const define_functions_1 = require("./utils/define-functions");
function addPausable(c, access, pausableFns) {
    c.addParent({
        name: 'Pausable',
        path: '@openzeppelin/contracts/utils/Pausable.sol',
    });
    for (const fn of pausableFns) {
        c.addModifier('whenNotPaused', fn);
    }
    addPauseFunctions(c, access);
}
function addPauseFunctions(c, access) {
    (0, set_access_control_1.requireAccessControl)(c, functions.pause, access, 'PAUSER', 'pauser');
    c.addFunctionCode('_pause();', functions.pause);
    (0, set_access_control_1.requireAccessControl)(c, functions.unpause, access, 'PAUSER', 'pauser');
    c.addFunctionCode('_unpause();', functions.unpause);
}
const functions = (0, define_functions_1.defineFunctions)({
    pause: {
        kind: 'public',
        args: [],
    },
    unpause: {
        kind: 'public',
        args: [],
    },
});
