// @openzeppelin/wizard/dist/governor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.numberPattern = exports.timelockOptions = exports.votesOptions = exports.defaults = void 0;
exports.printGovernor = printGovernor;
exports.isAccessControlRequired = isAccessControlRequired;
exports.buildGovernor = buildGovernor;
const common_functions_1 = require("./common-functions");
const common_options_1 = require("./common-options");
const contract_1 = require("./contract");
const error_1 = require("./error");
const set_access_control_1 = require("./set-access-control");
const print_1 = require("./print");
const set_info_1 = require("./set-info");
const set_upgradeable_1 = require("./set-upgradeable");
const define_functions_1 = require("./utils/define-functions");
const duration_1 = require("./utils/duration");
const set_clock_mode_1 = require("./set-clock-mode");
exports.defaults = {
    name: 'MyGovernor',
    delay: '1 day',
    period: '1 week',
    votes: 'erc20votes',
    clockMode: set_clock_mode_1.clockModeDefault,
    timelock: 'openzeppelin',
    blockTime: 12,
    decimals: 18,
    proposalThreshold: '0',
    quorumMode: 'percent',
    quorumPercent: 4,
    quorumAbsolute: '',
    storage: false,
    settings: true,
    access: common_options_1.defaults.access,
    upgradeable: common_options_1.defaults.upgradeable,
    info: common_options_1.defaults.info
};
exports.votesOptions = ['erc20votes', 'erc721votes'];
exports.timelockOptions = [false, 'openzeppelin', 'compound'];
function printGovernor(opts = exports.defaults) {
    return (0, print_1.printContract)(buildGovernor(opts));
}
function isAccessControlRequired(opts) {
    return opts.upgradeable === 'uups';
}
function withDefaults(opts) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    return {
        ...opts,
        ...(0, common_options_1.withCommonDefaults)(opts),
        decimals: (_a = opts.decimals) !== null && _a !== void 0 ? _a : exports.defaults.decimals,
        blockTime: opts.blockTime || exports.defaults.blockTime,
        quorumPercent: (_b = opts.quorumPercent) !== null && _b !== void 0 ? _b : exports.defaults.quorumPercent,
        quorumAbsolute: (_c = opts.quorumAbsolute) !== null && _c !== void 0 ? _c : exports.defaults.quorumAbsolute,
        proposalThreshold: opts.proposalThreshold || exports.defaults.proposalThreshold,
        settings: (_d = opts.settings) !== null && _d !== void 0 ? _d : exports.defaults.settings,
        storage: (_e = opts.storage) !== null && _e !== void 0 ? _e : exports.defaults.storage,
        quorumMode: (_f = opts.quorumMode) !== null && _f !== void 0 ? _f : exports.defaults.quorumMode,
        votes: (_g = opts.votes) !== null && _g !== void 0 ? _g : exports.defaults.votes,
        clockMode: (_h = opts.clockMode) !== null && _h !== void 0 ? _h : exports.defaults.clockMode,
        timelock: (_j = opts.timelock) !== null && _j !== void 0 ? _j : exports.defaults.timelock
    };
}
function buildGovernor(opts) {
    const allOpts = withDefaults(opts);
    const c = new contract_1.ContractBuilder(allOpts.name);
    validateDecimals(allOpts.decimals);
    addBase(c, allOpts);
    addSettings(c, allOpts);
    addCounting(c);
    addStorage(c, allOpts);
    addVotes(c);
    addQuorum(c, allOpts);
    addTimelock(c, allOpts);
    (0, set_access_control_1.setAccessControl)(c, allOpts.access);
    (0, set_upgradeable_1.setUpgradeable)(c, allOpts.upgradeable, allOpts.access);
    (0, set_info_1.setInfo)(c, allOpts.info);
    return c;
}
function addBase(c, { name }) {
    const Governor = {
        name: 'Governor',
        path: '@openzeppelin/contracts/governance/Governor.sol',
    };
    c.addParent(Governor, [name]);
    c.addOverride(Governor, functions.votingDelay);
    c.addOverride(Governor, functions.votingPeriod);
    c.addOverride(Governor, functions.quorum);
    c.addOverride(Governor, functions.state);
    c.addOverride(Governor, functions.propose);
    c.addOverride(Governor, functions.proposalNeedsQueuing);
    c.addOverride(Governor, functions.proposalThreshold);
    c.addOverride(Governor, functions._propose);
    c.addOverride(Governor, functions._queueOperations);
    c.addOverride(Governor, functions._executeOperations);
    c.addOverride(Governor, functions._cancel);
    c.addOverride(Governor, functions._executor);
    c.addOverride(Governor, common_functions_1.supportsInterface);
}
function addSettings(c, allOpts) {
    if (allOpts.settings) {
        const GovernorSettings = {
            name: 'GovernorSettings',
            path: '@openzeppelin/contracts/governance/extensions/GovernorSettings.sol',
        };
        c.addParent(GovernorSettings, [
            getVotingDelay(allOpts),
            getVotingPeriod(allOpts),
            { lit: getProposalThreshold(allOpts) },
        ]);
        c.addOverride(GovernorSettings, functions.votingDelay, 'view');
        c.addOverride(GovernorSettings, functions.votingPeriod, 'view');
        c.addOverride(GovernorSettings, functions.proposalThreshold, 'view');
    }
    else {
        setVotingParameters(c, allOpts);
        setProposalThreshold(c, allOpts);
    }
}
function getVotingDelay(opts) {
    try {
        if (opts.clockMode === 'timestamp') {
            return {
                lit: (0, duration_1.durationToTimestamp)(opts.delay),
            };
        }
        else {
            return {
                value: (0, duration_1.durationToBlocks)(opts.delay, opts.blockTime),
                note: opts.delay
            };
        }
    }
    catch (e) {
        if (e instanceof Error) {
            throw new error_1.OptionsError({
                delay: e.message,
            });
        }
        else {
            throw e;
        }
    }
}
function getVotingPeriod(opts) {
    try {
        if (opts.clockMode === 'timestamp') {
            return {
                lit: (0, duration_1.durationToTimestamp)(opts.period),
            };
        }
        else {
            return {
                value: (0, duration_1.durationToBlocks)(opts.period, opts.blockTime),
                note: opts.period
            };
        }
    }
    catch (e) {
        if (e instanceof Error) {
            throw new error_1.OptionsError({
                period: e.message,
            });
        }
        else {
            throw e;
        }
    }
}
function validateDecimals(decimals) {
    if (!/^\d+$/.test(decimals.toString())) {
        throw new error_1.OptionsError({
            decimals: 'Not a valid number',
        });
    }
}
function getProposalThreshold({ proposalThreshold, decimals, votes }) {
    if (!/^\d+$/.test(proposalThreshold)) {
        throw new error_1.OptionsError({
            proposalThreshold: 'Not a valid number',
        });
    }
    if (/^0+$/.test(proposalThreshold) || decimals === 0 || votes === 'erc721votes') {
        return proposalThreshold;
    }
    else {
        return `${proposalThreshold}e${decimals}`;
    }
}
function setVotingParameters(c, opts) {
    const delayBlocks = getVotingDelay(opts);
    if ('lit' in delayBlocks) {
        c.setFunctionBody([`return ${delayBlocks.lit};`], functions.votingDelay);
    }
    else {
        c.setFunctionBody([`return ${delayBlocks.value}; // ${delayBlocks.note}`], functions.votingDelay);
    }
    const periodBlocks = getVotingPeriod(opts);
    if ('lit' in periodBlocks) {
        c.setFunctionBody([`return ${periodBlocks.lit};`], functions.votingPeriod);
    }
    else {
        c.setFunctionBody([`return ${periodBlocks.value}; // ${periodBlocks.note}`], functions.votingPeriod);
    }
}
function setProposalThreshold(c, opts) {
    const threshold = getProposalThreshold(opts);
    const nonZeroThreshold = parseInt(threshold) !== 0;
    if (nonZeroThreshold) {
        c.setFunctionBody([`return ${threshold};`], functions.proposalThreshold);
    }
}
function addCounting(c) {
    c.addParent({
        name: 'GovernorCountingSimple',
        path: '@openzeppelin/contracts/governance/extensions/GovernorCountingSimple.sol',
    });
}
function addVotes(c) {
    const tokenArg = '_token';
    c.addImportOnly({
        name: 'IVotes',
        path: `@openzeppelin/contracts/governance/utils/IVotes.sol`,
        transpiled: false,
    });
    c.addConstructorArgument({
        type: {
            name: 'IVotes',
            transpiled: false,
        },
        name: tokenArg,
    });
    c.addParent({
        name: 'GovernorVotes',
        path: `@openzeppelin/contracts/governance/extensions/GovernorVotes.sol`,
    }, [{ lit: tokenArg }]);
}
exports.numberPattern = /^(?!$)(\d*)(?:\.(\d+))?(?:e(\d+))?$/;
function addQuorum(c, opts) {
    if (opts.quorumMode === 'percent') {
        if (opts.quorumPercent > 100) {
            throw new error_1.OptionsError({
                quorumPercent: 'Invalid percentage',
            });
        }
        let { quorumFractionNumerator, quorumFractionDenominator } = getQuorumFractionComponents(opts.quorumPercent);
        const GovernorVotesQuorumFraction = {
            name: 'GovernorVotesQuorumFraction',
            path: '@openzeppelin/contracts/governance/extensions/GovernorVotesQuorumFraction.sol',
        };
        if (quorumFractionDenominator !== undefined) {
            c.addOverride(GovernorVotesQuorumFraction, functions.quorumDenominator);
            c.setFunctionBody([
                `return ${quorumFractionDenominator};`
            ], functions.quorumDenominator, 'pure');
        }
        c.addParent(GovernorVotesQuorumFraction, [quorumFractionNumerator]);
        c.addOverride(GovernorVotesQuorumFraction, functions.quorum);
    }
    else if (opts.quorumMode === 'absolute') {
        if (!exports.numberPattern.test(opts.quorumAbsolute)) {
            throw new error_1.OptionsError({
                quorumAbsolute: 'Not a valid number',
            });
        }
        let returnStatement = (opts.decimals === 0 || opts.votes === 'erc721votes') ?
            `return ${opts.quorumAbsolute};` :
            `return ${opts.quorumAbsolute}e${opts.decimals};`;
        c.setFunctionBody([
            returnStatement,
        ], functions.quorum, 'pure');
    }
}
const timelockModules = {
    openzeppelin: {
        timelockType: {
            name: 'TimelockController',
            path: `@openzeppelin/contracts/governance/TimelockController.sol`,
        },
        timelockParent: {
            name: 'GovernorTimelockControl',
            path: `@openzeppelin/contracts/governance/extensions/GovernorTimelockControl.sol`,
        }
    },
    compound: {
        timelockType: {
            name: 'ICompoundTimelock',
            path: `@openzeppelin/contracts/vendor/compound/ICompoundTimelock.sol`,
            transpiled: false,
        },
        timelockParent: {
            name: 'GovernorTimelockCompound',
            path: `@openzeppelin/contracts/governance/extensions/GovernorTimelockCompound.sol`,
        }
    },
};
function getQuorumFractionComponents(quorumPercent) {
    let quorumFractionNumerator = quorumPercent;
    let quorumFractionDenominator = undefined;
    const quorumPercentSegments = quorumPercent.toString().split(".");
    if (quorumPercentSegments.length > 2) {
        throw new error_1.OptionsError({
            quorumPercent: 'Invalid percentage',
        });
    }
    else if (quorumPercentSegments.length == 2 && quorumPercentSegments[0] !== undefined && quorumPercentSegments[1] !== undefined) {
        quorumFractionNumerator = parseInt(quorumPercentSegments[0].concat(quorumPercentSegments[1]));
        const decimals = quorumPercentSegments[1].length;
        quorumFractionDenominator = '100';
        while (quorumFractionDenominator.length < decimals + 3) {
            quorumFractionDenominator += '0';
        }
    }
    return { quorumFractionNumerator, quorumFractionDenominator };
}
function addTimelock(c, { timelock }) {
    if (timelock === false) {
        return;
    }
    const timelockArg = '_timelock';
    const { timelockType, timelockParent } = timelockModules[timelock];
    c.addImportOnly(timelockType);
    c.addConstructorArgument({
        type: timelockType,
        name: timelockArg,
    });
    c.addParent(timelockParent, [{ lit: timelockArg }]);
    c.addOverride(timelockParent, functions._queueOperations);
    c.addOverride(timelockParent, functions._executeOperations);
    c.addOverride(timelockParent, functions._cancel);
    c.addOverride(timelockParent, functions._executor);
    c.addOverride(timelockParent, functions.state);
    c.addOverride(timelockParent, functions.proposalNeedsQueuing);
}
function addStorage(c, { storage }) {
    if (storage) {
        const GovernorStorage = {
            name: 'GovernorStorage',
            path: '@openzeppelin/contracts/governance/extensions/GovernorStorage.sol',
        };
        c.addParent(GovernorStorage);
        c.addOverride(GovernorStorage, functions._propose);
    }
}
const functions = (0, define_functions_1.defineFunctions)({
    votingDelay: {
        args: [],
        returns: ['uint256'],
        kind: 'public',
        mutability: 'pure',
    },
    votingPeriod: {
        args: [],
        returns: ['uint256'],
        kind: 'public',
        mutability: 'pure',
    },
    proposalThreshold: {
        args: [],
        returns: ['uint256'],
        kind: 'public',
        mutability: 'pure',
    },
    proposalNeedsQueuing: {
        args: [
            { name: 'proposalId', type: 'uint256' },
        ],
        returns: ['bool'],
        kind: 'public',
        mutability: 'view',
    },
    quorum: {
        args: [
            { name: 'blockNumber', type: 'uint256' },
        ],
        returns: ['uint256'],
        kind: 'public',
        mutability: 'view',
    },
    quorumDenominator: {
        args: [],
        returns: ['uint256'],
        kind: 'public',
        mutability: 'view',
    },
    propose: {
        args: [
            { name: 'targets', type: 'address[] memory' },
            { name: 'values', type: 'uint256[] memory' },
            { name: 'calldatas', type: 'bytes[] memory' },
            { name: 'description', type: 'string memory' },
        ],
        returns: ['uint256'],
        kind: 'public',
    },
    _propose: {
        args: [
            { name: 'targets', type: 'address[] memory' },
            { name: 'values', type: 'uint256[] memory' },
            { name: 'calldatas', type: 'bytes[] memory' },
            { name: 'description', type: 'string memory' },
            { name: 'proposer', type: 'address' },
        ],
        returns: ['uint256'],
        kind: 'internal',
    },
    _queueOperations: {
        args: [
            { name: 'proposalId', type: 'uint256' },
            { name: 'targets', type: 'address[] memory' },
            { name: 'values', type: 'uint256[] memory' },
            { name: 'calldatas', type: 'bytes[] memory' },
            { name: 'descriptionHash', type: 'bytes32' },
        ],
        kind: 'internal',
        returns: ['uint48'],
    },
    _executeOperations: {
        args: [
            { name: 'proposalId', type: 'uint256' },
            { name: 'targets', type: 'address[] memory' },
            { name: 'values', type: 'uint256[] memory' },
            { name: 'calldatas', type: 'bytes[] memory' },
            { name: 'descriptionHash', type: 'bytes32' },
        ],
        kind: 'internal',
    },
    _cancel: {
        args: [
            { name: 'targets', type: 'address[] memory' },
            { name: 'values', type: 'uint256[] memory' },
            { name: 'calldatas', type: 'bytes[] memory' },
            { name: 'descriptionHash', type: 'bytes32' },
        ],
        returns: ['uint256'],
        kind: 'internal',
    },
    state: {
        args: [
            { name: 'proposalId', type: 'uint256' },
        ],
        returns: ['ProposalState'],
        kind: 'public',
        mutability: 'view',
    },
    _executor: {
        args: [],
        returns: ['address'],
        kind: 'internal',
        mutability: 'view',
    },
});
