// @openzeppelin/wizard/dist/set-clock-mode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clockModeDefault = exports.clockModeOptions = void 0;
exports.setClockMode = setClockMode;
const define_functions_1 = require("./utils/define-functions");
exports.clockModeOptions = ['blocknumber', 'timestamp'];
exports.clockModeDefault = 'blocknumber';
const functions = (0, define_functions_1.defineFunctions)({
    clock: {
        kind: 'public',
        args: [],
        returns: ['uint48'],
        mutability: 'view',
    },
    CLOCK_MODE: {
        kind: 'public',
        args: [],
        returns: ['string memory'],
        mutability: 'pure',
    }
});
function setClockMode(c, parent, votes) {
    if (votes === 'timestamp') {
        c.addOverride(parent, functions.clock);
        c.setFunctionBody(['return uint48(block.timestamp);'], functions.clock);
        c.setFunctionComments(['// solhint-disable-next-line func-name-mixedcase'], functions.CLOCK_MODE);
        c.addOverride(parent, functions.CLOCK_MODE);
        c.setFunctionBody(['return "mode=timestamp";'], functions.CLOCK_MODE);
    }
}
