// @openzeppelin/wizard/dist/set-upgradeable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.upgradeableOptions = void 0;
exports.setUpgradeable = setUpgradeable;
const set_access_control_1 = require("./set-access-control");
const define_functions_1 = require("./utils/define-functions");
exports.upgradeableOptions = [false, 'transparent', 'uups'];
function setUpgradeable(c, upgradeable, access) {
    if (upgradeable === false) {
        return;
    }
    c.upgradeable = true;
    c.addParent({
        name: 'Initializable',
        path: '@openzeppelin/contracts/proxy/utils/Initializable.sol',
    });
    switch (upgradeable) {
        case 'transparent': break;
        case 'uups': {
            (0, set_access_control_1.requireAccessControl)(c, functions._authorizeUpgrade, access, 'UPGRADER', 'upgrader');
            const UUPSUpgradeable = {
                name: 'UUPSUpgradeable',
                path: '@openzeppelin/contracts/proxy/utils/UUPSUpgradeable.sol',
            };
            c.addParent(UUPSUpgradeable);
            c.addOverride(UUPSUpgradeable, functions._authorizeUpgrade);
            c.setFunctionBody([], functions._authorizeUpgrade);
            break;
        }
        default: {
            const _ = upgradeable;
            throw new Error('Unknown value for `upgradeable`');
        }
    }
}
const functions = (0, define_functions_1.defineFunctions)({
    _authorizeUpgrade: {
        args: [
            { name: 'newImplementation', type: 'address' },
        ],
        kind: 'internal',
    },
});
