// @openzeppelin/wizard/dist/generate/stablecoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateStablecoinOptions = generateStablecoinOptions;
const set_access_control_1 = require("../set-access-control");
const set_clock_mode_1 = require("../set-clock-mode");
const set_info_1 = require("../set-info");
const set_upgradeable_1 = require("../set-upgradeable");
const alternatives_1 = require("./alternatives");
const booleans = [true, false];
const blueprint = {
    name: ['MyStablecoin'],
    symbol: ['MST'],
    burnable: booleans,
    pausable: booleans,
    mintable: booleans,
    permit: booleans,
    limitations: [false, 'allowlist', 'blocklist'],
    votes: [...booleans, ...set_clock_mode_1.clockModeOptions],
    flashmint: booleans,
    premint: ['1'],
    custodian: booleans,
    access: set_access_control_1.accessOptions,
    upgradeable: set_upgradeable_1.upgradeableOptions,
    info: set_info_1.infoOptions,
};
function* generateStablecoinOptions() {
    for (const opts of (0, alternatives_1.generateAlternatives)(blueprint)) {
        yield { ...opts, upgradeable: false };
    }
}
